src/shared_realm.cpp

```diff
@@ -931,6 +950,84 @@ template Object Realm::resolve_thread_safe_reference(ThreadSafeReference<Object>
+static_assert(static_cast<int>(ComputedPrivileges::Delete) == static_cast<int>(sync::Privilege::Delete), "");
+static_assert(static_cast<int>(ComputedPrivileges::SetPermissions) == static_cast<int>(sync::Privilege::SetPermissions), "");
+static_assert(static_cast<int>(ComputedPrivileges::Query) == static_cast<int>(sync::Privilege::Query), "");
+static_assert(static_cast<int>(ComputedPrivileges::Create) == static_cast<int>(sync::Privilege::Create), "");
+static_assert(static_cast<int>(ComputedPrivileges::ModifySchema) == static_cast<int>(sync::Privilege::ModifySchema), "");
+
+bool Realm::init_permission_cache()
+{
+    verify_thread();
+
+    if (m_permissions_cache) {
+        // Rather than trying to track changes to permissions tables, just skip the caching
+        // entirely within write transactions for now
+        if (is_in_transaction())
+            m_permissions_cache->clear();
+        return true;
+    }
+
+    // Admin users bypass permissions checks outside of the logic in PermissionsCache
+    if (m_config.sync_config && m_config.sync_config->is_partial && !m_config.sync_config->user->is_admin()) {
+        m_permissions_cache = std::make_unique<sync::PermissionsCache>(read_group(), m_config.sync_config->user->identity());
+        return true;
+    }
+    return false;
+}
+
+void Realm::invalidate_permission_cache()
+{
+    if (m_permissions_cache)
+        m_permissions_cache->clear();
+}
+
+ComputedPrivileges Realm::get_privileges()
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    return static_cast<ComputedPrivileges>(m_permissions_cache->get_realm_privileges());
+}
+
+static uint8_t inherited_mask(uint32_t privileges)
+{
+    uint8_t mask = ~0;
+    if (!(privileges & sync::Privilege::Read))
+        mask = 0;
+    else if (!(privileges & sync::Privilege::Update))
+        mask = static_cast<uint8_t>(sync::Privilege::Read | sync::Privilege::Query);
+    return mask;
+}
+
+ComputedPrivileges Realm::get_privileges(StringData object_type)
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    auto privileges = inherited_mask(m_permissions_cache->get_realm_privileges())
+                    & m_permissions_cache->get_class_privileges(object_type);
+    return static_cast<ComputedPrivileges>(privileges);
+}
+
+ComputedPrivileges Realm::get_privileges(RowExpr row)
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+
+    auto& table = *row.get_table();
+    auto object_type = ObjectStore::object_type_for_table_name(table.get_name());
+    sync::GlobalID global_id{object_type, sync::object_id_for_row(read_group(), table, row.get_index())};
+    auto privileges = inherited_mask(m_permissions_cache->get_realm_privileges())
+                    & inherited_mask(m_permissions_cache->get_class_privileges(object_type))
+                    & m_permissions_cache->get_object_privileges(global_id);
```

Comment:
Why not just use `PermissionsCache::can()`?

Reply:
I pushed this implementation last week, but looks like it got overridden by a force push: https://github.com/realm/realm-object-store/pull/610/files#diff-55a15b0c9125ec5f7989470e6eb8b217R1018

Reply:
Damn. I wish we were a little less laissez faire about rebasing and force pushing. 🙁